airbyte-integrations/connectors/destination-bigquery/src/main/java/io/airbyte/integrations/destination/bigquery/typing_deduping/BigQuerySqlGenerator.java

```diff
@@ -0,0 +1,465 @@
+      return extractAndCast(column, chosenType);
+    } else if (airbyteType instanceof Struct) {
+      // We need to validate that the struct is actually a struct.
+      // Note that struct columns are actually nullable in two ways. For a column `foo`:
+      // {foo: null} and {} are both valid, and are both written to the final table as a SQL NULL (_not_ a
+      // JSON null).
+      // JSON_QUERY(JSON'{}', '$.foo') returns a SQL null.
+      // JSON_QUERY(JSON'{"foo": null}', '$.foo') returns a JSON null.
+      return new StringSubstitutor(Map.of("column_name", column.originalName())).replace(
+          """
+          CASE
+            WHEN JSON_QUERY(`_airbyte_data`, '$.${column_name}') IS NULL
+              OR JSON_TYPE(JSON_QUERY(`_airbyte_data`, '$.${column_name}')) != 'object'
+              THEN NULL
+            ELSE JSON_QUERY(`_airbyte_data`, '$.${column_name}')
+          END
+          """);
+    } else if (airbyteType instanceof Array) {
+      // Much like the Struct case above, arrays need special handling.
+      return new StringSubstitutor(Map.of("column_name", column.originalName())).replace(
+          """
+          CASE
+            WHEN JSON_QUERY(`_airbyte_data`, '$.${column_name}') IS NULL
+              OR JSON_TYPE(JSON_QUERY(`_airbyte_data`, '$.${column_name}')) != 'array'
+              THEN NULL
+            ELSE JSON_QUERY(`_airbyte_data`, '$.${column_name}')
+          END
+          """);
+    } else if (airbyteType instanceof UnsupportedOneOf || airbyteType == AirbyteProtocolType.UNKNOWN) {
+      // JSON_VALUE converts JSON types to native SQL types (int64, string, etc.)
+      // We use JSON_QUERY rather than JSON_VALUE so that we can extract a JSON-typed value.
+      // This is to avoid needing to convert the raw SQL type back into JSON.
+      return "JSON_QUERY(`_airbyte_data`, '$." + column.originalName() + "')";
+    } else {
+      final StandardSQLTypeName dialectType = toDialectType(airbyteType);
+      return "SAFE_CAST(JSON_VALUE(`_airbyte_data`, '$." + column.originalName() + "') as " + dialectType.name() + ")";
+    }
+  }
+
+  public StandardSQLTypeName toDialectType(final AirbyteProtocolType airbyteProtocolType) {
+    return switch (airbyteProtocolType) {
+      // TODO doublecheck these
+      case STRING -> StandardSQLTypeName.STRING;
+      case NUMBER -> StandardSQLTypeName.NUMERIC;
+      case INTEGER -> StandardSQLTypeName.INT64;
+      case BOOLEAN -> StandardSQLTypeName.BOOL;
+      case TIMESTAMP_WITH_TIMEZONE -> StandardSQLTypeName.TIMESTAMP;
+      case TIMESTAMP_WITHOUT_TIMEZONE -> StandardSQLTypeName.DATETIME;
+      case TIME_WITH_TIMEZONE -> StandardSQLTypeName.STRING;
+      case TIME_WITHOUT_TIMEZONE -> StandardSQLTypeName.TIME;
+      case DATE -> StandardSQLTypeName.DATE;
+      case UNKNOWN -> StandardSQLTypeName.JSON;
+    };
+  }
+
+  @Override
+  public String createTable(final StreamConfig stream, final String suffix) {
```

Comment:
I think this and `BigQueryUtils` is where the current logic lives? But its using the Bigquery java sdk rather than SQL

Reply:
sorry realizing this comment isn't clear - we had talked about colocating some of the common operations. Not necessarily right now, but maybe its a good idea colocate any "create table/namespace" logic

Reply:
mmhm. I don't have a strong opinion on how to make this happen - a few weeks back I was pretty dogmatic about (1) use sql for everything, (2) keep separate classes for "generate SQL" and "actually interact with the db", but I feel less strongly after actually seeing the code written

e.g. maybe we just merge sqlgenerator+destinationhandler? or kill off destinationhandler and add everything into some other existing class